File: 프로그래머스/2/160585. 혼자서 하는 틱택토/혼자서 하는 틱택토.java
public class Solution {

    public static int solution(String[] board) {
        int answer = -1;

        int oCount = 0;
        int xCount = 0;
        for(int i = 0; i < 3; ++i)
        {
            for(int j = 0; j < 3; ++j)
            {
                if (board[i].charAt(j) == 'O')
                {
                    ++oCount;
                }
                else if (board[i].charAt(j) == 'X')
                {
                    ++xCount;
                }
            }
        }
        
                
        if(oCount - xCount > 1)
            return 0;

        if (oCount == xCount) {
            if (check(board, 'O'))
                answer = 0;
            else
                answer = 1;
        } else if(oCount > xCount)
        {
            if (check(board,'X'))
                answer = 0;
            else
                answer = 1;
        }
        else
            answer = 0;

        return answer;

    }

    public static boolean check(String[] board, char ch)
    {
        for(int i = 0; i < 3; ++i)
        {
            boolean flag = true;
            boolean flag2 = true;

            for(int j = 0; j < 3; ++j)
            {
                if (board[i].charAt(j) != ch)
                {
                    flag = false;
                }

                if (board[j].charAt(i) != ch)
                {
                    flag2 = false;
                }
            }


            if (flag || flag2)
            {
                return true;
            }
        }

        if (board[0].charAt(0) == ch && 
            board[1].charAt(1) == ch && 
            board[2].charAt(2) == ch)
            return true;

        if (board[0].charAt(2) == ch && 
            board[1].charAt(1) == ch && 
            board[2].charAt(0) == ch)
            return true;

        return false;
    }
}

File: 프로그래머스/2/160585. 혼자서 하는 틱택토/혼자서 하는 틱택토.cpp
#include <string>
#include <vector>

using namespace std;

int Count(const vector<string>& board, const char& target)
{
    int count = 0;
    
    for(int i = 0; i < board.size(); ++i)
    {
        for(int j = 0; j < board.size(); ++j)
        {
            if (board[i][j] == target)
            {
                ++count;
            }
        }
    }
    
    return count;
}

bool CheckIsEnd(const vector<string>& board, const char& target)
{
    for(int i = 0; i < board.size(); ++i)
    {
        if (board[i][0] == target)
        {
            if (board[i][0] == board[i][1] && board[i][0] == board[i][2])
                return true;
        }

        if (board[0][i] == target)
        {
            if (board[0][i] == board[1][i] && board[0][i] == board[2][i])
                return true;
        }
    }
    
    if (board[0][0] == target && board[0][0] == board[1][1] && board[0][0] == board[2][2])
        return true;
    
    if (board[0][2] == target && board[0][2] == board[1][1] && board[0][2] == board[2][0])
        return true;
    
    return false;
}

int solution(vector<string> board) {
    int answer = -1;
    
    int OCount = Count(board, 'O');
    int XCount = Count(board, 'X');
    
    if (XCount > OCount || OCount - XCount > 1)
    {
        return 0;
    }
    
    if (OCount == XCount)
    {
        if (CheckIsEnd(board, 'O') == true)
            return 0;
        return 1;
    }
    
    if (OCount - XCount == 1)
    {
        if (CheckIsEnd(board, 'X') == true)
            return 0;
        
        return 1;
    }
    
    return 11234123;
}